### ISSUE TITLE
Problems with nfs & high i/o traffic

### ISSUE THREAD

**Issue by flowers9** [PacificBiosciences/FALCON#221]
I know nfs isn't the best way to share filesystems, but it's what we've got here.  When running falcon using multiple computers, during various points i/o traffic to the disk server can be high (although I do what I can to limit it), and attempts to open a file can time out.  There is a fairly simple fix (it's a kludge, but it works) of trying multiple times with a pause between tries.  I've modified Fopen in DB.c with
```c
  // in a cluster environment, nfs disks can fail to immediately display
  // files created on another machine; if the file isn't there, give it
  // a few times, with 30 seconds between them
  for (i = 0; i < 4 && f == NULL; ++i) {
      if ((f = fopen(name, newmode)) == NULL) {
          if (i < 3) {
              EPRINTF(EPLACE, "%s: Cannot open %s for '%s', retrying\n", Prog_Name, name, mode);
              sleep(30);
          } else {
              EPRINTF(EPLACE, "%s: Cannot open %s for '%s'\n", Prog_Name, name, mode);
          }
      }
  }
```
Perhaps not ideal, but it has been helping our runs complete without having to be restarted multiple times.  If you have access to a filesystem that doesn't have this issue, it shouldn't cause a delay, as the fopen() won't fail.  The only issue would be if you ran one of the daligner/dazzler programs by hand on a file that actually didn't exist, and it still generates an error message, allowing you to interrupt manually, so I don't see too much of an issue with time lost.

**Comment by pb-cdunn** [comment]
I'll read this more closely when I have time. But first, have you tried the tmp-dir option?

    use_tmpdir = True

We should document that, since it seems to be working reliably.

**Comment by flowers9** [comment]
That seems to have been introduced after 7/22/2015, which is when I last pulled the code.  As I have a lot of personal modifications at this point, I think I'll just stick with my solution, but thank you for the pointer.

**Comment by pb-cdunn** [comment]
We accept pull-requests pretty readily.

I see now that this is not about reducing I/O -- which is accomplished by `use_tmpdir`, very effectively according to some users -- but rather forcing NFS to sync. All you need is to call `opendir()`.

Do you know at what point in the workflow this is happening? Or for which executable(s)?

**Comment by flowers9** [comment]
It mainly seems to be a problem with LA4Falcon, during pread creation.  My suspicion is that it's fallout from LAmerge, which does a lot of i/o, as consensus generation tasks are sent off as merges are finished (so they compete traffic-wise with LAmerge jobs in any reasonable sized run).  I've set up my grid queues to limit LAmerge to one job per computer, but with 11 computers in the cluster, even that is a lot of stress on the disk server.

**Comment by flowers9** [comment]
It mainly seems to show up for LA4Falcon, during pread creation in stage 0.  These jobs get run in parallel with LAmerge tasks for reasonable sized runs, and LAmerge involves a lot of i/o traffic.  I've set up my grid queues to limit LAmerge jobs to one per machine, but with 11 machines in the cluster, it's still a high load on the disk server.

(As an aside, I've also modified daligner to store results on a local disk and do the first LAmerge step locally, which helps a little on the disk traffic.)

**Comment by pb-cdunn** [comment]
With the latest code, all LAsort and a full level of LAmerge occur with associated daligner jobs. And with `use_tmpdir` everything can occur in /tmp.

Like I said, reasonable pull-requests are welcome, and we continue to make improvements. Your code-base is diverging. That's your biggest problem.

**Comment by zyndagj** [comment]
Hey Chris,

I am using master branch of FALCON-integrate and have `use_tmpdir` in my configuration file, but LA4Falcon still pulls all data from my lustre file system. Most of the fc_consensus tasks are just sitting idle and no data was ever rsynced to /tmp on the compute nodes.

I did however get warnings that the daligner tasks ignored use_tmpdir, so the configuration is being parsed correctly. Below is my current configuration. Let me know if you can help at all.

Thanks,
Greg Zynda

**Example LA4Falcon Job**
```shell
$ cat c_00001.sub.sh
#!/bin/bash
set -vex

set -o pipefail
LA4Falcon -H12000  -fo ../raw_reads ../las_files/raw_reads.1.las | fc_consensus --output_dformat --min_id
t 0.70 --min_cov 6 --max_n_read 500 --n_core 16 --min_len_aln 500 --trim >| out.00001.fasta
```
**File Locations**
```shell
$ ls -lh las_files/                                                
total 0
lrwxrwxrwx 1 user group 27 Jan  8 11:40 raw_reads.10.las -> ../m_00010/raw_reads.10.las
lrwxrwxrwx 1 user group 27 Jan  8 11:40 raw_reads.11.las -> ../m_00011/raw_reads.11.las
lrwxrwxrwx 1 user group 27 Jan  8 11:42 raw_reads.12.las -> ../m_00012/raw_reads.12.las
lrwxrwxrwx 1 user group 27 Jan  8 11:40 raw_reads.13.las -> ../m_00013/raw_reads.13.las
```
**FALCON config**
```python
[General]
# list of files of the initial bas.h5/fasta files
input_fofn = input.fofn

# uncorrected input
input_type = raw
# corrected input
#input_type = preads

# The length cutoff used for seed reads used for initial mapping
length_cutoff = 12000

# The length cutoff used for seed reads usef for pre-assembly
length_cutoff_pr = 12000

# Use /tmp
use_tmpdir = True

job_type = SLURM
jobqueue = normal
allocation = Allocation
ncores = 16
sge_option_da  = -n %(ncores)s -t 04:00:00 -A %(allocation)s -p %(jobqueue)s
sge_option_la  = -n %(ncores)s -t 04:00:00 -A %(allocation)s -p %(jobqueue)s
sge_option_pda = -n %(ncores)s -t 04:00:00 -A %(allocation)s -p %(jobqueue)s
sge_option_pla = -n %(ncores)s -t 04:00:00 -A %(allocation)s -p %(jobqueue)s
sge_option_fc  = -n %(ncores)s -t 04:00:00 -A %(allocation)s -p %(jobqueue)s
sge_option_cns = -n %(ncores)s -t 04:00:00 -A %(allocation)s -p %(jobqueue)s

maxJobs = 30
pa_concurrent_jobs = %(maxJobs)s
ovlp_concurrent_jobs = %(maxJobs)s

pa_HPCdaligner_option = -k18 -w8 -h480 -t16 -M26 -dal128 -deg8 -H12000 -e0.75 -l4800
ovlp_HPCdaligner_option = -k24 -h1024 -t16 -M26 -dal128 -deg8 -e0.96 -l2500 -H12000

pa_DBsplit_option = -x700 -s400
ovlp_DBsplit_option = -x700 -s400

falcon_sense_option = --output_dformat --min_idt 0.70 --min_cov 6 --max_n_read 500 --n_core 16 --min_len_aln 500 --trim
cns_concurrent_jobs = %(maxJobs)s

overlap_filtering_setting = --max_diff 20 --max_cov 100 --min_cov 20 --bestn 10 --n_core 16
```

**Comment by pb-cdunn** [comment]
We dropped `use_tmpdir` recently -- not by choice, but only to make progress elsewhere. We'll add that back eventually.

> Most of the fc_consensus tasks are just sitting idle.

Is that the problem you're trying to solve? What happens if you run the LA4Falcon/fc_consensus script manually? Is this a new problem? (I mean, was it working before we dropped use_tmpdir?)

We use lustre and never see this problem. How many daligner jobs (job_*) and consensus jobs (m_*) do you have?

**Comment by zyndagj** [comment]
I have 51 blocks with 400Mbp (-s400) each and 30 compute nodes with 16 cores each.

```shell
$ ls -d job*/
job_0000/  job_0006/  job_000c/  job_0012/  job_0018/  job_001e/  job_0024/  job_002a/  job_0030/
job_0001/  job_0007/  job_000d/  job_0013/  job_0019/  job_001f/  job_0025/  job_002b/  job_0031/
job_0002/  job_0008/  job_000e/  job_0014/  job_001a/  job_0020/  job_0026/  job_002c/  job_0032/
job_0003/  job_0009/  job_000f/  job_0015/  job_001b/  job_0021/  job_0027/  job_002d/
job_0004/  job_000a/  job_0010/  job_0016/  job_001c/  job_0022/  job_0028/  job_002e/
job_0005/  job_000b/  job_0011/  job_0017/  job_001d/  job_0023/  job_0029/  job_002f/
$ ls -d job*/ | wc -l
51
```
**raw_reads.db**
```
files =        42
     119411 m150802_030215_42219_c100846802550000001823188212311524_s1_p0 m150802_030215_42219_c100846802
550000001823188212311524_s1_p0                                                                           
     137508 m150718_124843_42219_c100846622550000001823188212311561_s1_p0 m150718_124843_42219_c100846622
550000001823188212311561_s1_p0                                                                           
     154366 m150718_190748_42219_c100846622550000001823188212311562_s1_p0 m150718_190748_42219_c100846622
550000001823188212311562_s1_p0                                                                           
     258760 m150815_071435_42219_c100867602550000001823195203031664_s1_p0 m150815_071435_42219_c100867602
.
.
.
    2092101 m150801_080443_42219_c100846802550000001823188212311521_s1_p0 m150801_080443_42219_c100846802
550000001823188212311521_s1_p0                                                                           
    2190596 m150801_142340_42219_c100846802550000001823188212311522_s1_p0 m150801_142340_42219_c100846802
550000001823188212311522_s1_p0                                                                           
blocks =        51                                                                                       
size = 400000000 cutoff =       700 all = 0                                                              
```

**Comment by zyndagj** [comment]
Do you know the the patch that removed it? I'm also interested in trying /dev/shm since those tasks leave me with plenty of free memory so far.

**Comment by zyndagj** [comment]
I think I found the commit (72f821070d27e87f7fdd6338e005e883d5cab097), but it looks like falcon only used /tmp for the actual daligner step - not LA4Falcon.

**Comment by pb-cdunn** [comment]
I'm surprised that there are exactly 51 jobs for 51 blocks. Those are not 1:1.

Anyway, supporting use_tmpdir would take a bit of work. It also seemed to have zero value to some non-lustre users with large genomes, and we never have problems with lustre. (Note that this is an nfs thread, but you're using lustre.) So it's low priority now. But if you're sure it makes a big difference for you, I could work on it. I just need strong evidence so I can convince management.

Have you tried running the LA4Falcon/consensus script manually?

**Comment by zyndagj** [comment]
There were many more jobs before I increased the size setting. I'm also trying some of the settings @pb-jchin recommended for human assemblies.

I'll test running in /tmp, /dev/shm, and different stripe counts and report back with some results.

**Comment by pb-cdunn** [comment]
If you run in `/tmp`, you need `job_type=local` since cluster jobs will not see your files.

I do believe that running in /tmp can be a big improvement, but I've had trouble proving it. Moving a *single* job to /tmp has little measurable benefit. It's the sum of filesystem interactions that can be problematic.

**Comment by zyndagj** [comment]
After moving

* raw_reads.db
* .raw_reads.bps
* .raw_reads.idx

To /tmp, /dev/shm, and changing the stripe count and running c_*.sub.sh and looking at run stats with top I saw the following results:

## Lustre Stripe##
Increasing the stripe count from 2 to 6 led to no real gains and only utilized about 2 cores on average.
![stripe](https://4866bf29ccb5e57495fd3529d36d476604476f92.googledrive.com/host/0B1-hFGiszxLjQktFcjJHVEJ6Qm8/consensus_stripe_stats.jpg)
## /tmp ##
Running from /tmp was slightly faster with about 3 cores of usage on average with 16 threads.
![tmp](https://4866bf29ccb5e57495fd3529d36d476604476f92.googledrive.com/host/0B1-hFGiszxLjQktFcjJHVEJ6Qm8/consensus_tmp_stats.jpg)
## /dev/shm ##
All 10 of the workers were at 100% utilization in top, but it looks like there was some fine grained sleeping.
![shm](https://4866bf29ccb5e57495fd3529d36d476604476f92.googledrive.com/host/0B1-hFGiszxLjQktFcjJHVEJ6Qm8/consensus_shm_stats.jpg)

**Comment by pb-cdunn** [comment]
`/dev/shm` is shared memory, right? That could increase cpu utilization, but that doesn't mean the program will finish faster. But if it really does speed things up for you, I'd like to make that possible.

I haven't seen much speed-up for a given program by running in /tmp (versus lustre). But it's possible that running in /tmp helps *all* jobs -- including other people's jobs -- since it will use less network resources. That's really the effect I'd like to measure somehow.

Note that a VM might not have a true "local disk", so /tmp might offer zero value for VMs.
